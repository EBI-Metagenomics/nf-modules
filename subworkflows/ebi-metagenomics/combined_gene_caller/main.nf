
include { PYRODIGAL                } from '../../../modules/ebi-metagenomics/pyrodigal/main'
include { FRAGGENESCANRS           } from '../../../modules/ebi-metagenomics/fraggenescanrs/main'
include { COMBINEDGENECALLER_MERGE } from '../../../modules/ebi-metagenomics/combinedgenecaller/merge/main'

workflow COMBINED_GENE_CALLER {

    take:
    ch_assembly  // channel: [ val(meta), [ fasta ] ]
    ch_mask_file // channel: [ val(meta), [.out | .txt] ]

    main:

    ch_versions = Channel.empty()

    PYRODIGAL ( ch_assembly, "gff" )
    ch_versions = ch_versions.mix(PYRODIGAL.out.versions)

    FRAGGENESCANRS ( ch_assembly, params.fraggenescanrs_train_model )
    ch_versions = ch_versions.mix(FRAGGENESCANRS.out.versions)

    ch_mask_file = ch_mask_file ?: Channel.empty()

    ch_annotations = PYRODIGAL.out.annotations.join(
        PYRODIGAL.out.fna,
    ).join(
        PYRODIGAL.out.faa,
    ).join(
        FRAGGENESCANRS.out.gff,
    ).join(
        FRAGGENESCANRS.out.nucleotide_fasta,
    ).join(
        FRAGGENESCANRS.out.amino_acid_fasta,
    ).join(
        ch_mask_file, remainder: true
    )

    ch_merge = ch_annotations.map { meta, prodigal_sco, prodigal_nt_fasta, prodigal_aa_fasta, frag_gene_annot, frag_nt_fasta, frag_aa_fasta, mask ->
        return [
            meta,
            prodigal_sco,
            prodigal_nt_fasta,
            prodigal_aa_fasta,
            frag_gene_annot,
            frag_nt_fasta,
            frag_aa_fasta,
            mask ?: []
        ]
    }

    COMBINEDGENECALLER_MERGE( ch_merge )

    ch_versions = ch_versions.mix(COMBINEDGENECALLER_MERGE.out.versions)

    emit:
    faa      = COMBINEDGENECALLER_MERGE.out.faa  // channel: [ val(meta), [ faa ] ]
    ffn      = COMBINEDGENECALLER_MERGE.out.ffn  // channel: [ val(meta), [ ffn ] ]
    gff      = COMBINEDGENECALLER_MERGE.out.gff  // channel: [ val(meta), [ gff ] ]
    versions = ch_versions                       // channel: [ versions.yml ]
}
