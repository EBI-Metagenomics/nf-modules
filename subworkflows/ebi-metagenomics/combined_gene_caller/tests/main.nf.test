nextflow_workflow {

    name "Test Subworkflow COMBINED_GENE_CALLER"
    script "../main.nf"
    workflow "COMBINED_GENE_CALLER"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_ebimetagenomics"
    tag "subworkflows/combined_gene_caller"
    tag "combinedgenecaller/merge"
    tag "prodigal"
    tag "fraggenescan"

    test("cgc assembly fasta") {

        when {
            workflow {
                """
                input[0] = [
                    [ id:'test', single_end:false ],
                    file("${moduleDir}/tests/data/test_assembly.fasta.gz", checkIfExists: true)
                ]
                input[1] = channel.empty()
            """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.success },
                {
                    assert snapshot(
                        path(workflow.out.faa.get(0).get(1)).linesGzip[10..20],
                        path(workflow.out.gff.get(0).get(1)).linesGzip[0..10],
                        path(workflow.out.ffn.get(0).get(1)).linesGzip[10..20]
                    ).match()
                },
            )
        }
    }

    test("cgc assembly fasta and cmsearch mask file") {

        when {
            workflow {
                """
                input[0] = [
                    [ id:'test', single_end:false ],
                    file("${moduleDir}/tests/data/test_assembly.fasta.gz", checkIfExists: true)
                ]
                input[1] = channel.of([
                    [ id:'test', single_end:false ],
                    file("${moduleDir}/tests/data/test_tblout.deoverlapped", checkIfExists: true)
                ])
            """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                {
                    assert snapshot(
                        path(workflow.out.faa.get(0).get(1)).linesGzip[10..20],
                        path(workflow.out.gff.get(0).get(1)).linesGzip[0..10],
                        path(workflow.out.ffn.get(0).get(1)).linesGzip[10..20]
                    ).match()
                },
                {
                    def fasta = path(workflow.out.faa.get(0).get(1)).fasta
                    // // TODO: review this change (prodigal) ID=1_3 -> ID=2_3 (pyrodigal)
                    // assert "MGYA1_3 # 2446 # 2682 # 1 # ID=1_3;partial=00;start_type=TTG;rbs_motif=AGGAG;rbs_spacer=5-10bp;gc_cont=0.287" in fasta
                    // assert "MGYA1_3 # 2446 # 2682 # 1 # ID=2_3;partial=00;start_type=TTG;rbs_motif=AGGAG;rbs_spacer=5-10bp;gc_cont=0.287" in fasta
                    assert fasta.size() == 42
                }
            )
        }
    }

    test("cgc assembly fasta - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = [
                    [ id:'test', single_end:false ],
                    file("${moduleDir}/tests/data/test_assembly.fasta.gz", checkIfExists: true)
                ]
                input[1] = channel.empty()
            """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                // { assert snapshot(workflow.out).match() }
            )
        }
    }
}
