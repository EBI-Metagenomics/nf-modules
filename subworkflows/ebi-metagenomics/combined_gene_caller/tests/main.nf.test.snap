{
    "cgc assembly fasta and cmsearch mask file": {
        "content": [
            [
                ">MGYA1_4 # 2985 # 6581 # 1 # ID=2_4;partial=01;start_type=ATG;rbs_motif=None;rbs_spacer=None;gc_cont=0.604",
                "MTTKAPTFQQPLQSVVRLEGSAAMFEAHISGFPVPEVSWYRDGQVLSAATLPGVQISFSD",
                "GRAKLVISAVTAANSGRYTVQATNASGQATSTAELLVTAETAPPNFSQRLQSMTARQASR",
                "VRLDVRVTGINTPVVKFYRDGVEIQSSPDFIILHEGDLYSLIIAEAYPEDSGTYSVNATN",
                "SVGRATSTAELLVQGEDEAVPAKKTKTIVSTAQITQTTNQARVEKKIQAHFDARSDATAE",
                "MTDEAPKFLQPLQSVVALEGSAATEEARISGFPVPEVSWFRDGQVLSAAAPPGVQISFSD",
                "GRAKLMIPAVSTANSGRYSVRATNGAGQATSTAELLVTAETAMPNFTQRLQSTTARQGSQ",
                "VRFDIRVNGIPTPVVKFYRWGAEIQSSHDFQIIQDADLYSYIITEVYNEDSGTYSVNATN",
                "SVGRATSTCELIVQADEEIPAKKTKTIISAAQISQTRQARVEQRIESRFDARTIMSVEMT",
                "ESMQLPSWLENPLFGWRLLVDWLLGHIPEADIDQMRERAESRWTTGDKWMALAESHEAAA",
                "ADAGIHGSGVAATELAAAHRRMAQAWREQAEIQYSLAKQLYEGANEFEKTAWVIIISGIA"
            ],
            [
                "##gff-version 3",
                "MGYA1\tPyrodigal\tCDS\t1\t126\t.\t+\t.\tID=MGYA1_1",
                "MGYA1\tPyrodigal\tCDS\t1223\t1894\t.\t+\t.\tID=MGYA1_2",
                "MGYA1\tPyrodigal\tCDS\t2446\t2682\t.\t+\t.\tID=MGYA1_3",
                "MGYA1\tPyrodigal\tCDS\t2985\t6581\t.\t+\t.\tID=MGYA1_4",
                "MGYA2\tPyrodigal\tCDS\t347\t931\t.\t-\t.\tID=MGYA2_1",
                "MGYA2\tPyrodigal\tCDS\t959\t1894\t.\t-\t.\tID=MGYA2_2",
                "MGYA3\tPyrodigal\tCDS\t2\t616\t.\t-\t.\tID=MGYA3_1",
                "MGYA3\tPyrodigal\tCDS\t719\t3562\t.\t+\t.\tID=MGYA3_2",
                "MGYA4\tPyrodigal\tCDS\t229\t579\t.\t+\t.\tID=MGYA4_1",
                "MGYA4\tPyrodigal\tCDS\t4935\t5108\t.\t+\t.\tID=MGYA4_5"
            ],
            [
                "CAAGTCAATACAGAAGATATTCAAGCAGAACTACGACAAGAGATGCGTCAGGTTCAAGAT",
                "GAAAACAAAGAAATGTTACGTGAAGTGAGAGCATCACATGAGCATTATCAGAAGAGACAA",
                "AAACAATTATTTACAGGGATTGGATCGATGTTATTAGTCTTTATGCTGTTTGCTTTAATT",
                "ATGACCATAGGCAGTGACTTTATGAGTTTCTTGCATGTAGATATATTACAAAAAGCTATA",
                "GCAAGCAAAATCGAAGCTTCTGAAGGCTTTATGACATTTGTTTGGTATATTGCATACGGT",
                "TTACCTTATATATTCGCTATAGGACTATTTATTGGCTTATATGAATGGATTAGAGCACGG",
                "CTCCATGATTAA",
                ">MGYA1_3 # 2446 # 2682 # 1 # ID=2_3;partial=00;start_type=TTG;rbs_motif=AGGAG;rbs_spacer=5-10bp;gc_cont=0.287",
                "TTGAAATTTATAACTAAAATGTTAGATGGATTAGCTTTTATTTGTGCAGCATTATTATTT",
                "TTAAAGTATGCAGTTATGTCTGCCAATATGATGTTTGATTGGCATTTGCGTTGGTATTTC",
                "TTAGAAGATATTCCGCACATGGCAATTATATTATTCATCTTGATGTTTGTTTTTGCAGTT"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-18T14:45:20.721789"
    },
    "cgc assembly fasta": {
        "content": [
            [
                ">MGYA1_4 # 2985 # 6581 # 1 # ID=2_4;partial=01;start_type=ATG;rbs_motif=None;rbs_spacer=None;gc_cont=0.604",
                "MTTKAPTFQQPLQSVVRLEGSAAMFEAHISGFPVPEVSWYRDGQVLSAATLPGVQISFSD",
                "GRAKLVISAVTAANSGRYTVQATNASGQATSTAELLVTAETAPPNFSQRLQSMTARQASR",
                "VRLDVRVTGINTPVVKFYRDGVEIQSSPDFIILHEGDLYSLIIAEAYPEDSGTYSVNATN",
                "SVGRATSTAELLVQGEDEAVPAKKTKTIVSTAQITQTTNQARVEKKIQAHFDARSDATAE",
                "MTDEAPKFLQPLQSVVALEGSAATEEARISGFPVPEVSWFRDGQVLSAAAPPGVQISFSD",
                "GRAKLMIPAVSTANSGRYSVRATNGAGQATSTAELLVTAETAMPNFTQRLQSTTARQGSQ",
                "VRFDIRVNGIPTPVVKFYRWGAEIQSSHDFQIIQDADLYSYIITEVYNEDSGTYSVNATN",
                "SVGRATSTCELIVQADEEIPAKKTKTIISAAQISQTRQARVEQRIESRFDARTIMSVEMT",
                "ESMQLPSWLENPLFGWRLLVDWLLGHIPEADIDQMRERAESRWTTGDKWMALAESHEAAA",
                "ADAGIHGSGVAATELAAAHRRMAQAWREQAEIQYSLAKQLYEGANEFEKTAWVIIISGIA"
            ],
            [
                "##gff-version 3",
                "MGYA1\tPyrodigal\tCDS\t1\t126\t.\t+\t.\tID=MGYA1_1",
                "MGYA1\tPyrodigal\tCDS\t1223\t1894\t.\t+\t.\tID=MGYA1_2",
                "MGYA1\tPyrodigal\tCDS\t2446\t2682\t.\t+\t.\tID=MGYA1_3",
                "MGYA1\tPyrodigal\tCDS\t2985\t6581\t.\t+\t.\tID=MGYA1_4",
                "MGYA2\tPyrodigal\tCDS\t347\t931\t.\t-\t.\tID=MGYA2_1",
                "MGYA2\tPyrodigal\tCDS\t959\t1894\t.\t-\t.\tID=MGYA2_2",
                "MGYA3\tPyrodigal\tCDS\t2\t616\t.\t-\t.\tID=MGYA3_1",
                "MGYA3\tPyrodigal\tCDS\t719\t3562\t.\t+\t.\tID=MGYA3_2",
                "MGYA4\tPyrodigal\tCDS\t229\t579\t.\t+\t.\tID=MGYA4_1",
                "MGYA4\tPyrodigal\tCDS\t4935\t5108\t.\t+\t.\tID=MGYA4_5"
            ],
            [
                "CAAGTCAATACAGAAGATATTCAAGCAGAACTACGACAAGAGATGCGTCAGGTTCAAGAT",
                "GAAAACAAAGAAATGTTACGTGAAGTGAGAGCATCACATGAGCATTATCAGAAGAGACAA",
                "AAACAATTATTTACAGGGATTGGATCGATGTTATTAGTCTTTATGCTGTTTGCTTTAATT",
                "ATGACCATAGGCAGTGACTTTATGAGTTTCTTGCATGTAGATATATTACAAAAAGCTATA",
                "GCAAGCAAAATCGAAGCTTCTGAAGGCTTTATGACATTTGTTTGGTATATTGCATACGGT",
                "TTACCTTATATATTCGCTATAGGACTATTTATTGGCTTATATGAATGGATTAGAGCACGG",
                "CTCCATGATTAA",
                ">MGYA1_3 # 2446 # 2682 # 1 # ID=2_3;partial=00;start_type=TTG;rbs_motif=AGGAG;rbs_spacer=5-10bp;gc_cont=0.287",
                "TTGAAATTTATAACTAAAATGTTAGATGGATTAGCTTTTATTTGTGCAGCATTATTATTT",
                "TTAAAGTATGCAGTTATGTCTGCCAATATGATGTTTGATTGGCATTTGCGTTGGTATTTC",
                "TTAGAAGATATTCCGCACATGGCAATTATATTATTCATCTTGATGTTTGTTTTTGCAGTT"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-18T14:55:47.296979"
    }
}